fdbserver/DataDistribution.actor.cpp

```diff
@@ -2224,26 +2217,37 @@ ACTOR Future<Void> scheduleAuditStorageShardOnServer(Reference<DataDistributor>
 						self->remainingBudgetForAuditTasks[auditType].set(
 						    SERVER_KNOBS->CONCURRENT_AUDIT_TASK_COUNT_MAX - 1);
 					}
+					issueDoAuditCount++;
 					AuditStorageRequest req(audit->coreState.id, auditStates[i].range, auditType);
-					audit->actors.add(doAuditOnStorageServer(self, audit, ssi, req));
+					doAuditTasks.add(doAuditOnStorageServer(self, audit, ssi, req));
 				}
 			}
 			wait(delay(0.1));
 		}
+
+		wait(doAuditTasks.getResult());
 		TraceEvent(SevInfo, "DDScheduleAuditStorageShardOnServerEnd", self->ddId)
 		    .detail("ServerID", serverId)
 		    .detail("AuditID", audit->coreState.id)
 		    .detail("AuditType", auditType)
-		    .detail("TotalRanges", totalCount)
-		    .detail("TotalComplete", completedCount)
-		    .detail("CompleteRatio", completedCount * 1.0 / totalCount);
+		    .detail("IssuedDoAuditCount", issueDoAuditCount);
 
 	} catch (Error& e) {
-		TraceEvent(SevWarn, "DDScheduleAuditStorageShardOnServerError", self->ddId)
+		TraceEvent(SevInfo, "DDScheduleAuditStorageShardOnServerError", self->ddId)
 		    .errorUnsuppressed(e)
 		    .detail("AuditID", audit->coreState.id)
-		    .detail("AuditType", auditType);
-		audit->anyChildAuditFailed = true;
+		    .detail("AuditType", auditType)
+		    .detail("IssuedDoAuditCount", issueDoAuditCount);
+		if (e.code() == error_code_actor_cancelled) {
+			throw e;
+		} else if (e.code() == error_code_broken_promise) {
```

Comment:
Why is this one special? Can we merge this into the last `else {`?

Reply:
This is special only for ssshard audit, where DD goes through the list of SSes and issue audit task to each of the SSes. However, since we are doing throttling, there can be a big delay between the moment when DD reads the SS interface list and the moment when the DD interface is used. It is possible that SS gets removed in the middle (say the interface of this SS is invalid). To avoid repeatedly retrying for an invalid SS interface, `doAuditOnStorageServer` throws broken_promise if it is failed without a reply from SS. Then the scheduler captures this error and distinguish it as a signal to avoid repeatedly retry on this invalid interface. Instead, the schedule will trigger a global retry and re-read the entire SS list.